core/trino-main/src/main/java/io/trino/execution/scheduler/StageTaskSourceFactory.java

```diff
@@ -0,0 +1,705 @@
+        private final Optional<BucketNodeMap> bucketNodeMap;
+        private final Optional<CatalogName> catalogRequirement;
+
+        private boolean finished;
+        private boolean closed;
+
+        public static HashDistributionTaskSource create(
+                Session session,
+                PlanFragment fragment,
+                SplitSourceFactory splitSourceFactory,
+                Multimap<PlanFragmentId, ExchangeSourceHandle> exchangeSourceHandles,
+                int splitBatchSize,
+                LongConsumer getSplitTimeRecorder,
+                Optional<int[]> bucketToPartitionMap,
+                Optional<BucketNodeMap> bucketNodeMap)
+        {
+            checkArgument(bucketNodeMap.isPresent() || fragment.getPartitionedSources().isEmpty(), "bucketNodeMap is expected to be set");
+            Map<PlanNodeId, SplitSource> splitSources = splitSourceFactory.createSplitSources(session, fragment);
+            return new HashDistributionTaskSource(
+                    splitSources,
+                    getPartitionedExchangeSourceHandles(fragment, exchangeSourceHandles),
+                    getReplicatedExchangeSourceHandles(fragment, exchangeSourceHandles),
+                    splitBatchSize,
+                    getSplitTimeRecorder,
+                    bucketToPartitionMap,
+                    bucketNodeMap,
+                    fragment.getPartitioning().getConnectorId());
+        }
+
+        public HashDistributionTaskSource(
+                Map<PlanNodeId, SplitSource> splitSources,
+                Multimap<PlanNodeId, ExchangeSourceHandle> partitionedExchangeSourceHandles,
+                Multimap<PlanNodeId, ExchangeSourceHandle> replicatedExchangeSourceHandles,
+                int splitBatchSize,
+                LongConsumer getSplitTimeRecorder,
+                Optional<int[]> bucketToPartitionMap,
+                Optional<BucketNodeMap> bucketNodeMap,
+                Optional<CatalogName> catalogRequirement)
+        {
+            this.splitSources = ImmutableMap.copyOf(requireNonNull(splitSources, "splitSources is null"));
+            this.partitionedExchangeSourceHandles = ImmutableListMultimap.copyOf(requireNonNull(partitionedExchangeSourceHandles, "partitionedExchangeSourceHandles is null"));
+            this.replicatedExchangeSourceHandles = ImmutableListMultimap.copyOf(requireNonNull(replicatedExchangeSourceHandles, "replicatedExchangeSourceHandles is null"));
+            this.splitBatchSize = splitBatchSize;
+            this.getSplitTimeRecorder = requireNonNull(getSplitTimeRecorder, "getSplitTimeRecorder is null");
+            this.bucketToPartitionMap = requireNonNull(bucketToPartitionMap, "bucketToPartitionMap is null");
+            checkArgument(bucketToPartitionMap.isPresent() || partitionedExchangeSourceHandles.isEmpty() || splitSources.isEmpty(), "bucket to partition map is expected to be set when partitioned inputs are present");
```

Comment:
why do we have `|| splitSources.isEmpty()` here?

Reply:
Actually `bucketToPartitionMap` should always be present (with the current implementation) for hash distributed stages. It may not be the case in the future. I will drop the `Optional` for now. In the future we will refactor if needed.